### ISSUE TITLE
UPNP/NAT-PMP support

### ISSUE THREAD

**Issue by Sakooooo** [slskd/slskd#908]
Nicotine+ and SoulSeekQT have UPNP and NAT-PMP. could it be possible to have that as an option?

**Comment by Sakooooo** [comment]
alright ill try adding it in sometime maybe if i can

**Comment by pbogre** [comment]
i would also appreciate this feature since other services like transmission have built-in UPNP support and it makes the port-forwarding side of things a lot easier

**Comment by jpdillingham** [comment]
I should have clarified at the time, but I'm hesitant to add this because UPnP (and NAT-PMP from what I gather) is discouraged due to security concerns (here's one article: https://www.upguard.com/blog/what-is-upnp)

I'm not strictly opposed to adding the feature but it should definitely be opt-in and the documentation should warn users somehow.

**Comment by Sakooooo** [comment]
wait upnp is a security flaw!? dammit

**Comment by mathiascode** [comment]
> I should have clarified at the time, but I'm hesitant to add this because UPnP (and NAT-PMP from what I gather) is discouraged due to security concerns (here's one article: https://www.upguard.com/blog/what-is-upnp)
> 
> 
> 
> I'm not strictly opposed to adding the feature but it should definitely be opt-in and the documentation should warn users somehow.

NAT-PMP is a simpler protocol strictly focused on forwarding ports, as opposed to UPnP, which has access to more settings, so it should be safer in that aspect. In any case, I've enabled both by default in Nicotine+ in hopes of getting as many open ports on the Soulseek network as possible. If someone has disabled UPnP/NAT-PMP, the operation will silently fail anyway (with a log message).

It might of course be a good idea to encourage users to disable UPnP and manually forward ports instead. Perhaps slskd users would be more inclined to do so anyway.

**Comment by Sakooooo** [comment]
> It might of course be a good idea to encourage users to disable UPnP and manually forward ports instead. Perhaps slskd users would be more inclined to do so anyway.

just a reminder if anyone cant port forward cause they are behind  NAT check for another admin dashboard if you are using a wire from router to router for internet, that worked for me

**Comment by pbogre** [comment]
i have transmission inside a docker container, and that project using upnp, however i still had to do manual port forwarding through my router dashboard to allow incoming connections.

if there is any way to make upnp function correctly within a docker container i'd be happy to learn about it, both for slskd and other services that i run

**Comment by Sakooooo** [comment]
> i have transmission inside a docker container, and that project using upnp, however i still had to do manual port forwarding through my router dashboard to allow incoming connections.
> 
> 
> 
> if there is any way to make upnp function correctly within a docker container i'd be happy to learn about it, both for slskd and other services that i run

i dont think it might be possible had qbittorrent and that didn't show up on the upnp list
